### PR TITLE
Use custom format checker to validate date-time strings

### PR DESCRIPTION
**Pull request by kruai**
# Overview

This PR is being created to address [ESSNTL-1170](https://issues.redhat.com/browse/ESSNTL-1170).
Changes our date-time string validation to be more lenient, allowing time zone data to be omitted.
I had to remove the old implementation, because it was necessarily strict and didn't allow for that kind of customization. Instead, I've opted to redo this using a custom validator approach.

The original PR to add date-time validation is [here](https://github.com/RedHatInsights/insights-host-inventory/pull/968).

## PR Checklist

- [x] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [x] Tests: validate optimal/expected output
- [x] Tests: validate exceptions and failure scenarios
- [x] Tests: edge cases
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [x] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [x] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [x] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/validators.py [hunk 1]
```diff
@@ -45,6 +45,17 @@ def verify_mac_address_format(mac_address):
     return bool(pattern.match(mac_address))
 
 
+@draft4_format_checker.checks("date-time")
+def is_custom_date(val):
+    if not val:
+        return True
+    try:
+        match = re.match(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|\+\d{2}:\d{2})?$", val)
```
Comment (RonnyPfannschmidt) [inline app/validators.py:53]: is this the isoformat, would it be ok to just try datetime.fromisoformat here ?

Comment (kruai) [reply app/validators.py:53]: Could be, though this comment caused me to realize that we use `dateutil.parser.isoparse()` elsewhere in the code, which the `fromisoformat` docs say is preferable. I'll change it up to use that.

Comment (kruai) [reply app/validators.py]: I swapped it out for isoparse, and it's certainly simpler!

**Review by thearifismail — DISMISSED** [review]
LGTM

**Comment by thearifismail** [comment]
Approved it because all tests pass BUT pytest shutdown ends with following message.  The message is reported when  "test_duplicate_hosts.py" is run.

File: app/validators.py [hunk 2]
```diff
@@ -45,6 +46,17 @@ def verify_mac_address_format(mac_address):
     return bool(pattern.match(mac_address))
 
 
+@draft4_format_checker.checks("date-time")
+def is_custom_date(val):
+    if not val:
```
[the remark this answers is not in the document]
Comment (kruai) [reply app/validators.py:51]: Fair point

**Review by RonnyPfannschmidt — APPROVED** [review]

**Review by thearifismail — APPROVED** [review]

**Pull request merged**